lighthouse-cli/test/smokehouse/concurrent-mapper.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/**
+ * A class that maintains a concurrency pool to coordinate many jobs that should
+ * only be run `concurrencyLimit` at a time.
+ * API inspired by http://bluebirdjs.com/docs/api/promise.map.html, but
+ * independent callers of `concurrentMap()` share the same concurrency limit.
+ */
+class ConcurrentMapper {
+  constructor() {
+    /** @type {Set<Promise<any>>} */
+    this._promisePool = new Set();
+  }
+
+  /**
+   * Runs callbackfn on `values` in parallel, at a max of `concurrencyLimit` at
+   * a time across all callers on this instance. Resolves to an array of the
+   * results (for each caller separately) or rejects with the first rejected
+   * result.
+   * @template T, U
+   * @param {Array<T>} values
+   * @param {(value: T, index: number, array: Array<T>) => Promise<U>} callbackfn
+   * @param {number} concurrencyLimit
+   * @return {Promise<Array<U>>}
+   */
+  async concurrentMap(values, callbackfn, concurrencyLimit) {
```

Comment:
any reason all this is a class, instead of just thing one function? we wouldn't need to pass around an instance of this if that were the case.

Reply:
> any reason all this is a class, instead of just thing one function? we wouldn't need to pass around an instance of this if that were the case.

yeah, the stored promise pool lets you do
```js
const concurrentMapper = new ConcurrentMapper();
const resultA = concurrentMapper.concurrentMap(valuesA, mapFn, 5);
const resultB = concurrentMapper.concurrentMap(valuesB, mapFn, 2);
const results = [...await resultA, ...await resultB];
```
and have A jobs only run if there are 5 or fewer (A or B jobs) running at once and B jobs when 2 or fewer (A or B jobs), while still keeping the individual results separate and in order.

I just realized there's a bug in the current version I'll have to fix, though :)